torch/csrc/distributed/c10d/init.cpp

```diff
@@ -975,18 +976,21 @@ that adds a prefix to each key inserted to the store.
 
           .def(
               "broadcast",
-              &::c10d::ProcessGroup::broadcast,
+              [](const c10::intrusive_ptr<::c10d::ProcessGroup>& self,
```

Comment:
If I understand correctly, this change introduces some discrepancy between the C++ and Python Process Group APIs right? In C++ you have to explicitly call `dispatch_broadcast()` while in Python calling `pg.broadcast()` already goes through dispatcher. Not sure whether we want this sort of behavioral difference. This might also have implication on our existing high-level APIs (i.e. DDP, FSDP, & co.)

Reply:
Yeah you are going to want ProcessGroup::broadcast to also go through the dispatcher, the manual changes below are going to be very error prone otherwise.

Reply:
I think we still need to have the real ProcessGroup::broadcast to actually invoke nccl's broadcast. And then ones that route to the dispatcher.

Putting everything under ProcessGroup is an option, but then there are still two sets of APIs. I guess I can having ProcessGroup::broadcast being the dispatcher's one. And then introduce ProcessGroup::broadcast_impl to be the one that calls into nccl.

Reply:
@kumpera Could you illustrate what's the benefits by doing so?

Reply:
@alanwaketan can you summarize the outcome of this discussion? was there a conclusion?

Reply:
The conclusion is that we have separate APIs for C++ and Python. In Python, the pybind11 is used directly to route to the dispatcher. In C++, since the ProcessGroup's APIs are also the interface for 3rd party to adopt, we can't change that. We have to introduce the ops::xxx API. It's error prone. Given we shouldn't write complicated C++ stuff anymore, it should be okay?